src/passes/GUFA.cpp

```diff
@@ -0,0 +1,303 @@
+  bool canReplaceWithUnreachable(Expression* curr) {
+    if (!canRemoveStructurally(curr)) {
+      return false;
+    }
+    EffectAnalyzer effects(getPassOptions(), *getModule(), curr);
+    // Ignore a trap, as the unreachable replacement would trap too.
+    effects.trap = false;
+    return !effects.hasUnremovableSideEffects();
+  }
+
+  void visitExpression(Expression* curr) {
+    // Skip things we can't improve in any way.
+    auto type = curr->type;
+    if (type == Type::unreachable || type == Type::none ||
+        Properties::isConstantExpression(curr)) {
+      return;
+    }
+
+    if (type.isTuple()) {
+      // TODO: tuple types.
+      return;
+    }
+
+    if (type.isRef() && getTypeSystem() != TypeSystem::Nominal) {
+      // Without nominal typing we can't analyze subtypes, so we cannot infer
+      // anything about refs.
+      return;
+    }
+
+    // Ok, this is an interesting location that we might optimize. See what the
+    // oracle says is possible there.
+    auto contents = oracle.getContents(ExpressionLocation{curr, 0});
+
+    auto& options = getPassOptions();
+    auto& wasm = *getModule();
+    Builder builder(wasm);
+
+    auto replaceWithUnreachable = [&]() {
+      if (canReplaceWithUnreachable(curr)) {
+        replaceCurrent(getDroppedChildrenAndAppend(
+          curr, wasm, options, builder.makeUnreachable()));
+      } else {
+        // We can't remove this, but we can at least put an unreachable
+        // right after it.
+        replaceCurrent(builder.makeSequence(builder.makeDrop(curr),
+                                            builder.makeUnreachable()));
+      }
+      optimized = true;
+    };
+
+    if (contents.getType() == Type::unreachable) {
+      // This cannot contain any possible value at all. It must be unreachable
+      // code.
+      replaceWithUnreachable();
+      return;
+    }
+
+    // This is reachable. Check if we can emit something optimized for it.
+    // TODO: can we handle more general things here too?
+    if (!contents.canMakeExpression()) {
+      return;
+    }
+
+    if (contents.isNull() && curr->type.isNullable()) {
+      // Null values are all identical, so just fix up the type here (the null's
+      // type might not fit in this expression).
+      contents =
+        PossibleContents::literal(Literal::makeNull(curr->type.getHeapType()));
+
+      // Note that if curr's type is *not* nullable, then the code will trap at
+      // runtime (the null must arrive through a cast that will trap). We handle
+      // that below, so we don't need to think about it here.
+
+      // TODO: would emitting a more specific null be useful when valid?
+    }
+
+    auto* c = contents.makeExpression(wasm);
+
+    // We can only place the constant value here if it has the right type. For
+    // example, a block may return (ref any), that is, not allow a null, but in
+    // practice only a null may flow there if it goes through casts that will
+    // trap at runtime.
```

Comment:
Good question. I believe right now it does not, since we don't have full optimization/analysis of `ref.as` etc. Once we have all that, we might be able to assert here on the type being valid. I'll add a TODO.